src/scalar.jl

```diff
@@ -52,7 +53,144 @@ transform(::Identity, x::Real) = x
 
 transform_and_logjac(::Identity, x::Real) = x, zero(x)
 
-inverse(::Identity, x::Real) = x
+inverse(::Identity, x::Number) = x
+
+inverse_and_logjac(::Identity, x::Real) = x, zero(x)
+
+####
+#### elementary scalar transforms
+####
+
+"""
+$(TYPEDEF)
+
+Exponential transformation `x ↦ eˣ`. Maps from all reals to the positive reals.
+"""
+struct TVExp <: ScalarTransform 
+end
+transform(::TVExp, x::Real) = exp(x)
+transform_and_logjac(t::TVExp, x::Real) = transform(t, x), x
+
+function inverse(::TVExp, x::Number)
+    @argcheck x > 0 DomainError
+    log(x)
+end
+inverse_and_logjac(t::TVExp, x::Number) = inverse(t, x), log(1/x)
+
+"""
+$(TYPEDEF)
+
+Logistic transformation `x ↦ logit(x)`. Maps from all reals to (0, 1).
+"""
+struct TVLogistic <: ScalarTransform
+end
+transform(::TVLogistic, x::Real) = logistic(x)
+transform_and_logjac(t::TVLogistic, x::Real) = transform(t, x), logistic_logjac(x)
+
+function inverse(::TVLogistic, x::Number)
+    @argcheck 0 < x < 1 DomainError
+    logit(x)
+end
+inverse_and_logjac(t::TVLogistic, x::Number) = inverse(t, x), logit_logjac(x)
+
+"""
+$(TYPEDEF)
+
+Shift transformation `x ↦ x + shift`. 
+"""
+struct TVShift{T <: Real} <: ScalarTransform
+    shift::T
+end
+transform(t::TVShift, x::Real) = x + t.shift
+transform_and_logjac(t::TVShift, x::Real) = transform(t, x), zero(x)
+
+inverse(t::TVShift, x::Number) = x - t.shift
+inverse_and_logjac(t::TVShift, x::Number) = inverse(t, x), zero(x)
+
+"""
+$(TYPEDEF)
+
+Scale transformation `x ↦ scale * x`.
+"""
+struct TVScale{T} <: ScalarTransform
+    scale::T
+    function TVScale{T}(scale::T) where {T}
+        @argcheck scale > zero(scale) DomainError
+        new(scale)
+    end
+end
+TVScale(scale::T) where {T} = TVScale{T}(scale)
+
+transform(t::TVScale, x::Real) = t.scale * x
+transform_and_logjac(t::TVScale{T}, x::Real) where {T<:Real} = transform(t, x), log(t.scale) 
+
+inverse(t::TVScale, x::Number) = x / t.scale
+inverse_and_logjac(t::TVScale{T}, x::Number) where {T<:Real} = inverse(t, x), -log(t.scale)
+
+"""
+$(TYPEDEF)
+
+Negative transformation `x ↦ -x`.
+"""
+struct TVNeg <: ScalarTransform
+end
+
+transform(::TVNeg, x::Real) = -x
+transform_and_logjac(t::TVNeg, x::Real) = transform(t, x), zero(x)
+
+inverse(::TVNeg, x::Number) = -x
+inverse_and_logjac(::TVNeg, x::Number) = -x, zero(x)
+
+####
+#### composite scalar transforms
+####
+"""
+$(TYPEDEF)
+
+A composite scalar transformation, i.e. a sequence of scalar transformations.
+The component transforms can be accessed indexing into the `CompositeScalarTransform`, 
+as in `ct[i]` or `ct[end]`.
+"""
+struct CompositeScalarTransform{Ts <: Tuple} <: ScalarTransform
+    transforms::Ts
```

Comment:
I'm not sure about exact trade-offs - but noteably Julia itself uses a different design for function composition:
```suggestion
    outer_transform::T1
    inner_transform::T2
```
Presumably this would also simplify the definitions of `transform` and `inverse` below.

Reply:
This is a clear case of my just not knowing about more clever algorithms, with recursion like this. But I do like the idea of the transform being a tuple rather than a nested data structure, since it feels easier to reason about (e.g. how to recursively apply from the bottom of a nested struct)

Reply:
Personally I am OK with a `Tuple`, having a preference for flat structures.